File: src/main/java/de/photon/aacadditionpro/util/visibility/PlayerVisibility.java
package de.photon.aacadditionpro.util.visibility;

import org.bukkit.entity.Player;

public interface PlayerVisibility
{
    // In 1.18 there is a new draft api to handle this directly.
    PlayerVisibility INSTANCE = new LegacyPlayerVisibility();

    /**
     * This method will fully hide the toBeHidden {@link Player} from the observer {@link Player}
     */
    void fullyHidePlayer(Player observer, Player toBeHidden);

    /**
     * This method will hide the equipment of the hideEquipment {@link Player} from the observer {@link Player}
     */
    void hideEquipment(Player observer, Player hideEquipment);

    /**
     * This method will fully reveal the toBeRevealed {@link Player} from the observer {@link Player}
     */
    void revealPlayer(Player observer, Player toBeRevealed);

    void enable();

    void disable();
}

File: src/main/java/de/photon/aacadditionpro/util/visibility/PlayerVisibility.kt
package de.photon.aacadditionpro.util.visibility

import org.bukkit.entity.Player

object PlayerVisibility {
    private val equipmentHider: PlayerInformationHider = PlayerEquipmentHider()
    private val playerHider: PlayerInformationHider = PlayerHider()

    /**
     * This method will fully hide the toBeHidden [Player] from the observer [Player]
     */
    fun fullyHidePlayer(observer: Player?, toBeHidden: Player?) {
        playerHider.hidePlayer(observer, toBeHidden)
        equipmentHider.revealPlayer(observer, toBeHidden)
    }

    /**
     * This method will hide the equipment of the hideEquipment [Player] from the observer [Player]
     */
    fun hideEquipment(observer: Player?, hideEquipment: Player?) {
        equipmentHider.hidePlayer(observer, hideEquipment)
        playerHider.revealPlayer(observer, hideEquipment)
    }

    /**
     * This method will fully reveal the toBeRevealed [Player] from the observer [Player]
     */
    fun revealPlayer(observer: Player?, toBeRevealed: Player?) {
        playerHider.revealPlayer(observer, toBeRevealed)
        equipmentHider.revealPlayer(observer, toBeRevealed)
    }

    fun enable() {
        equipmentHider.registerListeners()
        playerHider.registerListeners()
    }

    fun disable() {
        equipmentHider.unregisterListeners()
        equipmentHider.clear()
        playerHider.unregisterListeners()
        playerHider.clear()
    }
}